presto-spi/src/main/java/com/facebook/presto/spi/relation/LambdaDefinitionExpression.java

```diff
@@ -111,4 +115,55 @@ private static void checkArgument(boolean condition, String message, Object... m
             throw new IllegalArgumentException(format(message, messageArgs));
         }
     }
+    private static class CanonicalizeExpression
+            implements RowExpressionVisitor<String, Void>
+{
+        private final Map<String, String> canonicalizedArguments = new HashMap<>();
+
+        public CanonicalizeExpression(List<String> arguments, List<Type> argumentTypes)
+        {
+            for (int i = 0; i < arguments.size(); i++) {
+                canonicalizedArguments.put(arguments.get(i), format("%s_%d", argumentTypes.get(i).toString(), i));
+            }
+        }
+
+        @Override
+        public String visitCall(CallExpression call, Void context)
+        {
+            return format("%s(%s)", call.getDisplayName(), String.join(", ", call.getArguments().stream().map(e -> e.accept(this, null)).collect(Collectors.toList())));
+        }
+
+        @Override
+        public String visitInputReference(InputReferenceExpression reference, Void context)
+        {
+            return reference.toString();
+        }
+
+        @Override
+        public String visitConstant(ConstantExpression literal, Void context)
+        {
+            return format("%d", literal.hashCode());
```

Comment:
`literal.toString()` doesn't work very well here because it would just write string literals as `Slice[...]`. I want to preserve the `equals` trait of literals. Better suggestions are welcome.

Reply:
This is too dangerous, isn't it? int32 hash collision is pretty high.

Reply:
I thought about this, but if there's a collision does that mean equals and map operations would give bogus results as well? Doesn't feel very dangerous after thinking about that. Still, I don't like this either, so if there's better alternatives, I'd take them.

Reply:
One idea is to use the JSON representation for `ConstantExpression ` (this actually also applies to other `RowExpression` here ! )

However it might be difficult to inject `JsonCodec<RowExpression>` here. If we look at `ConstantExpression#getValueBlock`, it actually calls to `Utils#nativeValueToBlock`:

Otherwise, doing an `instanceof Slice` might be OK... We did this in a couple of place, such as `LiteralEncoder`. And here we are "encoding constant literal" :) 

https://github.com/prestodb/presto/blob/f43e4441f46e56e6816bca0cfc8b66473a9c3b21/presto-main/src/main/java/com/facebook/presto/sql/planner/LiteralEncoder.java#L235-L241

Reply:
> but if there's a collision does that mean equals and map operations would give bogus results as well? 

If I understand correctly the answer is yes right? Since the `equals` implementation essentially compares the `argumentTypes` and `canonicalizedBody` string?:
```java
return Objects.equals(argumentTypes, that.argumentTypes) &&
        Objects.equals(canonicalizedBody, that.canonicalizedBody);
```